cpp/arcae/shared_fcntl_mutex.h

```diff
@@ -0,0 +1,160 @@
+#error Posix fcntl support required
+#else
+#include <fcntl.h>
+#include <unistd.h>
+#endif
+
+#include <arrow/result.h>
+#include <arrow/status.h>
+#include <arrow/util/logging.h>
+
+using namespace std::literals;
+
+namespace arcae {
+
+using FcntlLockType = decltype(flock::l_type);
+
+// Base locking class
+class BaseSharedFcntlMutex {
+ public:
+  virtual arrow::Status lock() = 0;
+  virtual arrow::Status lock_shared() = 0;
+  virtual arrow::Result<bool> try_lock() = 0;
+  virtual arrow::Result<bool> try_lock_shared() = 0;
+
+  virtual void unlock() = 0;
+  virtual void unlock_shared() = 0;
+  virtual ~BaseSharedFcntlMutex() {};
+};
+
+// Noop class, mostly for use with in memory tables
+class NullSharedFcntlMutex : public BaseSharedFcntlMutex {
+ public:
+  virtual arrow::Status lock() override { return arrow::Status::OK(); }
+  virtual arrow::Status lock_shared() override { return arrow::Status::OK(); }
+
+  virtual arrow::Result<bool> try_lock() override { return true; }
+  virtual arrow::Result<bool> try_lock_shared() override { return true; }
+
+  virtual void unlock() override {}
+  virtual void unlock_shared() override {};
+};
+
+// Two part lock guarding access to a CASA table across multiple processes
+//
+// It's interface is somewhat similar to C++ standard library mutexes
+// but methods may return arrow::{Status,Result} instead.
+//
+// It combines:
+//
+// 1. A per-process lock that provides multiple-reader single-writer (MRSW)
+//    access to an underlying
+// 2. fcntl process lock that coordinates multiple-reader single-writer (MRSW)
+//    access across processes.
+//
+// (1) is implemented using a std::shared_mutex. By it's nature this restricts
+// access to the underlying fcntl lock to a single thread in the case of writes.
+// Further care must be taken to restrict access of multiple readers
+// to the underlying fcntl lock: This is accomplished with a std::mutex
+// and a reader counter.
+class SharedFcntlMutex : public BaseSharedFcntlMutex,
+                         std::enable_shared_from_this<SharedFcntlMutex> {
+ private:
+  int fd_;
+  std::string lock_filename_;
+  // Multiple-reader Single-writer mutex
+  // Guards access at the process level
+  std::shared_mutex mutex_;
+  // Number of concurrent readers
+  int reader_counts_;
+  // Guards access to the fcntl lock
+  // This is only needed for reads as
+  // mutex_ will only ever allow one writer
+  std::mutex fcntl_read_mutex_;
+
+ public:
+  struct LockInfo {
+    FcntlLockType lock_type;
+    pid_t pid;
+  };
+
+ public:
+  // Primary factory function for creating an instance
+  static arrow::Result<std::shared_ptr<SharedFcntlMutex>> Create(
+      std::string_view lock_filename = "");
+
+  // Disable copies and moves
+  SharedFcntlMutex(const SharedFcntlMutex&) = delete;
+  SharedFcntlMutex(SharedFcntlMutex&& rhs) = delete;
+  SharedFcntlMutex& operator=(const SharedFcntlMutex&) = delete;
+  SharedFcntlMutex& operator=(SharedFcntlMutex&& rhs) = delete;
```

Comment:
Curiously, it seems like std::shared_mutex may be movable? But I don't see how that would work if you move a mutex out from under the lock guard that's locking it, so disabling move makes sense.

Reply:
It also enforces the use o f`SharedFcntlMutx::Create` to always wrap the mutex in a `shared_ptr` (for reasons discussed in another comment)